Validation/HLTrigger/plugins/HLTGenValSource.cc

```diff
@@ -0,0 +1,480 @@
+#include "Validation/HLTrigger/interface/HLTGenValHistCollPath.h"
+
+// DQMStore
+#include "DQMServices/Core/interface/DQMStore.h"
+#include "DQMServices/Core/interface/DQMEDAnalyzer.h"
+
+// trigger Event
+#include "HLTrigger/HLTcore/interface/HLTConfigProvider.h"
+#include "DataFormats/HLTReco/interface/TriggerEvent.h"
+
+// object that can be a GenJet, GenParticle or energy sum
+#include "Validation/HLTrigger/interface/HLTGenValObject.h"
+
+class HLTGenValSource : public DQMEDAnalyzer {
+public:
+  explicit HLTGenValSource(const edm::ParameterSet&);
+  ~HLTGenValSource() override = default;
+  HLTGenValSource(const HLTGenValSource&) = delete;
+  HLTGenValSource& operator=(const HLTGenValSource&) = delete;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void analyze(const edm::Event&, const edm::EventSetup&) override;
+  void bookHistograms(DQMStore::IBooker&, edm::Run const& run, edm::EventSetup const& c) override;
+  void dqmBeginRun(const edm::Run&, const edm::EventSetup&) override;
+
+  // functions to get correct object collection for chosen object type
+  std::vector<HLTGenValObject> getObjectCollection(const edm::Event&);
+  std::vector<HLTGenValObject> getGenParticles(const edm::Event&);
+  reco::GenParticle getLastCopyPreFSR(reco::GenParticle part);
+  reco::GenParticle getLastCopy(reco::GenParticle part);
+
+  // ----------member data ---------------------------
+
+  // tokens to get collections
+  const edm::EDGetTokenT<reco::GenParticleCollection> genParticleToken_;
+  const edm::EDGetTokenT<reco::GenMETCollection> genMETToken_;
+  const edm::EDGetTokenT<reco::GenJetCollection> ak4genJetToken_;
+  const edm::EDGetTokenT<reco::GenJetCollection> ak8genJetToken_;
+  const edm::EDGetTokenT<trigger::TriggerEvent> trigEventToken_;
+
+  // config strings/Psets
+  std::string objType_;
+  std::string dirName_;
+  std::vector<edm::ParameterSet> histConfigs_;
+  std::vector<edm::ParameterSet> histConfigs2D_;
+  std::vector<edm::ParameterSet> binnings_;
+  std::string hltProcessName_;
+
+  // constructing the info string, which will be written to the output file for display of information in the GUI
+  // the string will have a JSON formating, thus starting here with the opening bracket, which will be close directly before saving to the root file
+  std::string infoString_ = "{";
+
+  // histogram collection per path
+  std::vector<HLTGenValHistCollPath> collectionPath_;
+
+  // HLT config provider/getter
+  HLTConfigProvider hltConfig_;
+
+  // some miscellaneous member variables
+  std::vector<std::string> hltPathsToCheck_;
+  std::vector<std::string> hltPaths;
+  std::vector<std::string> hltPathSpecificCuts;
+  double dR2limit_;
+  bool doOnlyLastFilter_;
+};
+
+HLTGenValSource::HLTGenValSource(const edm::ParameterSet& iConfig)
+    : genParticleToken_(consumes<reco::GenParticleCollection>(
+          iConfig.getParameterSet("inputCollections").getParameter<edm::InputTag>("genParticles"))),
```

Comment:
(more for @cms-sw/dqm-l2) I had the impression that (at least) all new DQM code would use untracked parameters.

Reply:
This being new code, I'd also suggest being coherent to the prescription and move all parameters to be untracked, unless @cms-sw/dqm-l2 has a different opinion on it